### PR TITLE
Adds testdrive expiration

### PR DESCRIPTION
**Pull request by denibertovic**

### CODE REVIEW

**Comment by denibertovic** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;WcN6RrRM&#x2F;357-16-implement-test-drive-disable-expiration-feature"> (16) Implement test drive disable &#x2F; expiration feature</a>

**Comment by denibertovic** [comment]
@bezidejni this is very much a "first pass/needs testing" thingy. 

The UX could be better. It definitely needs to be tested on an actual cluster cause my devstack setup had some very very weird, but unrelated, issues that prevent from testing this properly. Also it might be worth styling the header.html expiration notice part..not sure.

**Comment by natea** [comment]
@denibertovic were you building this against Dogwood or Eucalyptus? Also, did you assign it to someone to do the code review?

**Comment by natea** [comment]
@denibertovic will this also block the user from accessing their course in Studio, or is it only on the LMS side? I didn't see any files touched in /cms which is why I'm asking.

**Comment by natea** [comment]
@denibertovic for the UI/styling in the header, could you ask @grozdanowski to take a look, and help to improve the look-n-feel of the expiration messaging? Also, the message that appears when the user's testdrive has expired.

**Comment by natea** [comment]
@denibertovic it appears that you're making these changes directly to the core edx-platform codebase, which could have implications when new versions of Open edX are released.

Did you talk to @bezidejni about the possibility of wrapping up this functionality as a standalone app, perhaps using his boilerplate template as a starting point? https://github.com/appsembler/edx-app-boilerplate

File: lms/templates/header.html [hunk 1]
```diff
@@ -8,6 +8,18 @@
 is_microsite = microsite.is_request_in_microsite()
 %>
 
+<% testdrive_expiration_enforced = user.profile.is_testdrive_expiration_enforced() %>
```
Comment (bezidejni) [inline lms/templates/header.html]: @denibertovic This throws an exception if the user is not logged in because an AnonymousUser doesn't have a profile attached to it. Btw, I don't think there's a need for that method on the UserProfile model because it's a global setting, and not a per-user setting. Here in the template you can just directly do `testdrive_expiration_enforced = settings.ENFORCE_TESTDRIVE_EXPIRATION`

File: lms/djangoapps/courseware/access.py [hunk 2]
```diff
@@ -102,6 +103,10 @@ def has_access(user, action, obj, course_key=None):
     if not user:
         user = AnonymousUser()
 
+    # We only check for testdrive expiration for actual users
+    if ((not user.is_anonymous()) and user.profile.is_testdrive_expired()):
+        return TestDriveExpiredError()
```
Comment (bezidejni) [inline lms/djangoapps/courseware/access.py]: Where is this caught/handled?

**Comment by bezidejni** [comment]
@denibertovic I see that it shows a message to the user that the trial has expired, but does it block the user from accessing the Studio/course in Studio somewhere?

**Comment by denibertovic** [comment]
Hey @natea  see replies below.

> were you building this against Dogwood or Eucalyptus? Also, did you assign it to someone to do the code review?

It's based off of our testdrive branch and that's on dogwood. And yes I've pinged @bezidejni for review. 

> will this also block the user from accessing their course in Studio, or is it only on the LMS side? 
> I didn't see any files touched in /cms which is why I'm asking.

The change was added to the core permission logic that's used in all of edx. My understanding is that it affects studio as well and that's what my testing showed. We should however definitely test it on a staging server.

> for the UI/styling in the header, could you ask @grozdanowski to take a look, and help to improve the > look-n-feel of the expiration messaging? Also, the message that appears when the user's testdrive has expired.

Will do.  But that's probably the last step once we're sure everything works.

> it appears that you're making these changes directly to the core edx-platform codebase, which could > have implications when new versions of Open edX are released.

I didn't find a straightforward way of doing this any more modular than it already is. The reality of the situation is that we're maintaining a fork and merging upstream will be a process either way.

File: lms/templates/header.html (hunk 1 above)
Comment (denibertovic) [inline lms/templates/header.html]: Nice catch. This was mostly a hack because of mako. I'll change this.

File: lms/djangoapps/courseware/access.py (hunk 2 above)
Comment (denibertovic) [inline lms/djangoapps/courseware/access.py]: It's handled where has_access is called. It's just checking the return value to be False and therefore refuses to render the content. I'm not sure where the user message is displayed though.

**Comment by denibertovic** [comment]
@bezidejni @natea Well, it seesm what I was seeing on studio was not that my restrictions were working but rather my dev env being borked. I pushed new changes that restrict studio access as well. (and show the header).

@bezidejni the only thing I'm concerned about now is what happens if we a new student enrolls into a course? Ie. creates a new account? I don't know if that's disabled on the testdrive but in my local env I can register a new user and it will have user.date_joined set in a way that it could continue using the testdrive.

**Comment by bezidejni** [comment]
@denibertovic I did some testing on devstack and the header notification works as intended, blocking studio access too. The only thing that we'll have to change (or remove) is checking  access permissions in the LMS part because if the test drive has expired, the user should still have access to his course in LMS, it should be blocked in Studio only if I'm not mistaken (cc @natea )

As for your question, even if a user somehow manages to enroll into that course, he or she still wouldn't be a staff/instructor user by default and have access to Studio which is the main thing we want to block the test drive expires. 

Once that LMS part is sorted, I think this is good to go.

**Comment by denibertovic** [comment]
@abeals  @natea is what Filip described accurate in terms of what we want to do?

**Comment by natea** [comment]
@denibertovic I think restricting their access to Studio is an adequate action to take when their test-drive expires. It will prevent them from making any content changes to their course. 

However, if someone knows they only have 14 days to build a course, it's possible that they could get it all done in that time frame, and start inviting students to the course on the LMS side, and have access to the instructor dashboard to actually run the course. 

I guess we can monitor for that kind of misuse of the test-drive, and if we see it being a problem, we can add another restriction to block their access to their course itself.

The advantage of keeping the course available for read-only purposes, is that champions can use the course as a way to get stakeholder buy-in / approval.

**Comment by natea** [comment]
@denibertovic @bezidejni could one of you guys take some screenshots that shows the whole workflow for the user, and the messages that are shown to them before and after their trial expires? I'm having a hard time providing valuable feedback without actually seeing the deliverable.

**Comment by bezidejni** [comment]
@natea If you think they can build a course in that time (and we decide to leave the LMS access open), we can then just reduce the length of the trial period to something less. But I think those cases are really edge cases which we can handle on an individual basis.

Here's some screenshots:
dashboard___appsembler_academy
dashboard___appsembler_academy

Please let us know what text should be there and anything else you'd like to be changed. Obiously, Matej will style that message before we go live. /cc @denibertovic @grozdanowski

**Comment by denibertovic** [comment]
@bezidejni I've added an additional variable to check for LMS restrictions specifically (instead of removing it completely ... in case we need it later). @grozdanowski  can you take a look at how to style the header? Perhaps we should wait til we have the exact text.

**Comment by denibertovic** [comment]
**IN SUMMARY:**


**This PR is waiting on:**

- Exact text for the error messages that currently read:
  - <span>Your TESTDRIVE has <strong>expired</strong>! Please email info@appsembler.com.</span>
  - <span>TESTDRIVE EXPIRES IN: ${days_til_testdrive_expires} days</span>
- Styling of the above

**Comment by natea** [comment]
> But I think those cases are really edge cases which we can handle on an individual basis.

@bezidejni how *would* we handle the edge case in which a user is requesting an extension of their test-drive beyond the 14 day trial limit? From @denibertovic 's code, it appears that the 14 day number is hardcoded, and there doesn't seem to be an easy way to give someone more time.

From the old Docker container Launcher service, it was quite common for us to extend a prospective customer's test-drive container and this even led to closing some deals that might not have happened, had we had a strict cut off after the 72 hrs.

**Comment by natea** [comment]
@denibertovic I'm not sure about having the CTA be to email info@appsembler.com. That seems very old school. At the very least we should send them to a Hubspot contact form, so we can track that correspondence. Maybe we want to make a special contact form for test-drive users, in which we prompt them for more info than what they filled out when they signed up for the test-drive.

Regarding the styling, could we have the background color be green when they have 14 days, orange after it hits 7 days, and red when they only have 3 days left? That way, it visually reminds them the urgency more than just a number counting down. /cc @grozdanowski

**Comment by denibertovic** [comment]
@natea we can go in the database manually and change the "date_joined" date do give them additional time.

**Comment by denibertovic** [comment]
(the reason I'm suggesting going in the database is because that field isn't changeable from the django admin I don't think....but if I'm wrong then it's more simple.)

**Comment by bezidejni** [comment]
It is changeable in django admin, I had to change it to do some testing.

**Comment by denibertovic** [comment]
Still need a decision for the CTA and text in general. And need @grozdanowski to style it.

**Comment by natea** [comment]
The thought occurred to me that if the test-drive user invites co-workers to enroll in the course or co-author the course, won't they have a different expiration date than the original user, because the expiration date is not tied to the date at which the test-drive was launched, but the date at which you create an account?

This could cause a lot of confusion whereby the person who originally authored the course no longer has access to modify it, but the users they invited can still access it.

I think we need to be consistent and have the expiration tied to the course itself, and not the user who created it. In other words, we're checking the "Date joined" value for the user, but we should be checking the "Course created" value to determine when the course should expire.

File: common/djangoapps/student/models.py [hunk 3]
```diff
@@ -384,6 +384,17 @@ def country_cache_key_name(cls, user_id):
         """
         return cls.PROFILE_COUNTRY_CACHE_KEY.format(user_id=user_id)
 
+    def is_testdrive_expired(self):
+        """Helper function that checks whether a given user's testdrive is expiried.
+        It will only enfore this check if ENFORCE_TESTDRIVE_EXPIRATION is set to True (default)."""
+        if settings.ENFORCE_TESTDRIVE_EXPIRATION:
+            return timezone.now().date() > (self.user.date_joined + timedelta(days=settings.TESTDRIVE_TRIAL_DAYS)).date()
```
Comment (natea) [inline common/djangoapps/student/models.py]: This should not be checking the user.date_joined value, but the course creation date value, since different users could have joined the site at different times, but the course should expire at the same time for all of them.

**Comment by denibertovic** [comment]
@natea I've thought about this. I've even mentioned it in the comments above. But since we're just disallowing access to studio and we've disallowed invites for staff users I figured it shouldn't be a problem. @bezidejni can you confirm that we only allow invites for students but not for staff users?

File: common/djangoapps/student/models.py (hunk 3 above)
Comment (bezidejni) [inline common/djangoapps/student/models.py]: @natea The user is created at the same time as the course, with maybe a few seconds of difference so it doesn't really matter.

**Comment by bezidejni** [comment]
@denibertovic Will check and get back to you.

**Comment by natea** [comment]
@bezidejni did you check whether the expiration notice only appears to staff users? 

@denibertovic are you saying that the test-drive user cannot invite other staff users to manage the course? this might limit the usefulness of the trial, since that is a key feature that test-drive users want to experience - collaborative authoring of courses. Can they also not invite instructors, or just authors?

**Comment by natea** [comment]
@grozdanowski bonus if you can make the background color where the expiration notice appears change color depending on how much time is left on their trial. ;)

Green - 7-14 days remaining
Yellow/Orange - 6 or less days remaining
Red - 0 days remaining (trial expired!)

**Comment by natea** [comment]
Is there anything else you or Matej need to merge this? For now, Aaron and
I decided that it's fine to send them to the /contact page on our website,
and we'll later change this to a dedicated Hubspot form for expired
test-drive users.

**Comment by denibertovic** [comment]
@natea @abeals I think we have everything. cc @grozdanowski

**Comment by denibertovic** [comment]
@bezidejni can you check if this looks right to you? If so let me know and I'll squash the commits before merging and then we can deploy (I'll need  your help for that).

**Comment by bezidejni** [comment]
@denibertovic I tried it and it looks good to me! Merge away!

**Comment by denibertovic** [comment]
Squashed commits. Merging.
